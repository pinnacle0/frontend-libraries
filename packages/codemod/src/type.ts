import type {ParserOptions} from "@babel/parser";
import type {Options as GenerateOptions, print} from "recast";
import type {ASTNode, builders, visit} from "ast-types";

export interface Toolkit {
    parse: (source: string) => ASTNode;
    builders: typeof builders;
    visit: typeof visit;
    generate: typeof print;
}

export interface CreateToolkitOptions {
    parse?: ParserOptions | undefined;
    generate?: GenerateOptions | undefined;
}

export type {NodePath} from "ast-types/lib/node-path";

export type Transform = (source: string, toolkit: Toolkit) => string | void;

export const Codemod = ["use-react-hook-from-web-ui", "use-react-util-from-web-ui", "object-freeze-to-class-static"] as const;

export type Codemod = typeof Codemod[number];
