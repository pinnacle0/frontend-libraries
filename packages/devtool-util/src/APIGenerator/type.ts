export interface APIDefinition {
    services: ServiceDefinition[];
    types: TypeDefinition[];
}

export interface ServiceDefinition {
    name: string;
    operations: ServiceOperation[];
}

export interface ServiceOperation {
    name: string;
    method: "DELETE" | "GET" | "POST" | "PUT";
    path: string;
    pathParams: {name: string; type: string}[];
    responseType: string;
    requestType: null | string;
    optional: boolean;
    deprecated: boolean;
}

export interface TypeDefinition {
    name: string;
    type: TypeDefinitionType;
    fields: TypeDefinitionField[] | null;
    enumConstants: TypeEnumConstant[] | null;
}

export type TypeDefinitionType = "enum" | "bean";

export type JavaType = "String" | "Boolean" | "Integer" | "Long" | "Double" | "BigDecimal" | "ZonedDateTime" | "LocalDate" | "LocalDateTime" | "LocalTime" | "List" | "Map";

export interface TypeDefinitionField {
    name: string;
    type: JavaType | string;
    typeParams: (JavaType | string)[] | null;
    constraints: TypeDefinitionFieldConstraints;
}

export interface TypeDefinitionFieldConstraints {
    notNull: boolean | null;
    notBlank: boolean | null;
    min: number | null;
    max: number | null;
    size: {min: number; max: number} | null;
    pattern: string | null;
}

export interface TypeEnumConstant {
    name: string;
    value: string;
}

export interface IgnoreType {
    enum?: {[key: string]: string[]};
    interface?: {[key: string]: string[]};
}

export interface PlatformConfig {
    ajaxFunction: string;
    ajaxFunctionImportStatement: string;
    typeFileImportPath: string;
}
export interface APIGeneratorOptions {
    metadataEndpointURL: string;
    typeFilePath: string;
    serviceFolderPath: string;
    ignoreType?: IgnoreType;
    platformConfig: PlatformConfig;
}

export type PlatformSpecificAPIGeneratorOptions = Omit<APIGeneratorOptions, "platformConfig">;
