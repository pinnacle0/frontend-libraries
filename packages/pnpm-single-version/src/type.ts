import type {nameVerFromPkgSnapshot} from "@pnpm/lockfile-utils";

export interface CheckerOptions {
    includes: string[];
    /**
     * default: true
     * In CI env, exit with status code 1 when checker found multiple version on specified dependencies.
     */
    failOnCI?: boolean;
}

export interface LoggerType {
    error(error: Error): void;
    message(value: string): void;
}

export type PackageInfo = ReturnType<typeof nameVerFromPkgSnapshot>;
