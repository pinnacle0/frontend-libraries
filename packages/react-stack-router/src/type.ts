import type React from "react";
import type {To} from "history";
import type {RouteProps} from "./component/Route";
import type {ScreenTransitionType} from "./stackRouter/screenTransition";
import type {ScreenHooks} from "./stackRouter/screen";

export type HistoryState = Record<string, any>;

export interface PushOption extends ScreenHooks {
    transition?: ScreenTransitionType;
    state?: HistoryState;
}

export interface Router {
    Root: React.ComponentType<React.PropsWithChildren>;
    Route: React.ComponentType<RouteProps>;
    push: (to: To, option?: PushOption) => void;
    pop: () => void;
    replace: (to: To, state?: Record<string, any>) => void;
    reset: () => void;
}
