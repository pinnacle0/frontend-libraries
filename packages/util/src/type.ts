import type React from "react";

export type ArrayElement<ArrayType extends readonly unknown[]> = ArrayType[number];

export type PickOptional<T> = Pick<T, {[K in keyof T]-?: {} extends {[P in K]: T[K]} ? K : never}[keyof T]>;
export type PickNonNullable<T> = Pick<T, NonNullableKeys<T>>;

export type KeysOfType<T, ExpectedValueType> = {[P in keyof T]: T[P] extends ExpectedValueType ? P : never}[keyof T];
export type NullableKeys<T> = {[K in keyof T]: T[K] extends NonNullable<T[K]> ? never : K}[keyof T];
export type NonNullableKeys<T> = {[K in keyof T]: T[K] extends NonNullable<T[K]> ? K : never}[keyof T];

export type MarkAsNonNullable<T, K extends keyof T> = Omit<T, K> & {[P in K]: NonNullable<T[P]>};
/**
 * @desc Return an interface where optionals are not allowed
 */
export type MarkAsRequired<T, K extends keyof T> = Omit<T, K> & Required<Pick<T, K>>;
export type MarkAsOptional<T, K extends keyof T> = Omit<T, K> & Partial<Pick<T, K>>;
export type MarkAsNullable<T, K extends keyof T> = Omit<T, K> & {[P in K]: T[P] | null};

/**
 * @desc Return an interface where optional, nullable and undefinable are not allowed
 */
export type StrictRequired<T> = {[K in keyof T]-?: Exclude<T[K], null | undefined>};

// TODO/Alvis: removed later, with eslint rules, web-ui internal
export type SafeReactChild = React.ReactChild | boolean | null;
export type SafeReactChildren = SafeReactChild | SafeReactChild[];

export interface ControlledFormValue<T> {
    value: T;
    onChange: (newValue: T) => void;
}
