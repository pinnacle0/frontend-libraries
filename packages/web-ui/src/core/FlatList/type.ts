import type React from "react";

export type Boundary = "top" | "bottom";

export type Gap =
    | number
    | {
          top?: number;
          bottom?: number;
          left?: number;
          right?: number;
      };

export interface FlatListItemProps<T> {
    data: T;
    index: number;
}

export interface FlatListProps<T> {
    /**
     * List of data to show in the list
     */
    data: T[];
    /**
     * Specify key for each `item` using Object key of type `T`
     */
    rowKey: keyof T | "index";
    /**
     * A component render for each item
     */
    renderItem: React.ComponentType<FlatListItemProps<T>>;
    /**
     * Set this to true when waiting for new data by pull down refresh
     */
    refreshing?: boolean;
    /**
     * Set this to true when waiting for new data by pull down refresh
     */
    loading?: boolean;
    /**
     * If Provided, `Refresh` component will be loaded into the FlatList, you can also specify your own Refresh component using `refresh` props
     */
    onPullDownRefresh?: () => void;
    /**
     * If Provided, `Footer` component will be loaded into the FlatList, you can also specify your own Refresh component using `footer` props
     */
    onPullUpLoading?: () => void;
    /**
     * If Provided, Element will be shown when `data` is a empty array
     */
    emptyPlaceholder?: string | React.ReactElement;
    /**
     * if Provided, default `Footer` component will be replaced
     */
    Footer?: React.ComponentType<{loading: boolean}>;
    /**
     * `refresh` component will be shown when `refreshing` is equal `true`
     * if Provided, default `Refresh` component will be replaced
     */
    Refresh?: React.ComponentType<{loading: boolean}>;
    /**
     * `pullUpMessage` will be show in `Refresh` only if `onPullDownRefresh` is defined
     * if Provided, this message will replace the pull down message in default `Refresh`
     * if `Refresh` is provided, this props will be ignored
     */
    pullDownMessage?: string;
    /**
     * `pullUpMessage` will be show in `Footer` only if `onPullUpLoading` is defined
     * if Provided, this message will replace end of list message in default `Footer`
     * if `Footer` is provided, this props will be ignored
     */
    pullUpMessage?: string;
    /**
     * if Provided, this message will replace end of list message in default `Footer`
     * if `Footer` is provided, this props will be ignored
     */
    endOfListMessage?: string;
    /**
     * Set margin for each item
     */
    gap?: Gap;
}
