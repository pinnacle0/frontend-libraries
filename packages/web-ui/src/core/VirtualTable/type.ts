export type ColumnFixedPosition = "left" | "right";

export interface StickyPosition {
    value: number;
    isLast: boolean;
}

/**
 * Similar usage of Antd Table but only support partial features: fixed columns, row selection, on row click
 */
export type VirtualTableRowSelection<RowType extends object> = {
    width: number;
    selectedRowKeys: React.Key[];
    onChange: (selectedRowKeys: React.Key[], selectedRows: RowType[]) => void;
    /**
     * Can only sticky in left
     */
    fixed?: boolean;
    isDisabled?: (data: RowType, rowIndex: number) => boolean;
    isSelectAllDisabled?: boolean;
    /**
     * Attention:
     * If title is provided, the select all checkbox wil be overridden
     */
    title?: React.ReactNode | string | number;
};

export type VirtualTableColumn<RowType extends object> = {
    title: React.ReactNode | string | number;
    width: number;
    /**
     * Attention:
     * If renderData return null, the corresponding table cell will not render
     */
    renderData: (record: RowType, rowIndex: number) => React.ReactNode | undefined;
    align?: "left" | "right" | "center";
    display?: "default" | "hidden";
    fixed?: "left" | "right";
    /**
     * Attention:
     * The overridden cell should return null in renderData props:
     * e.g. [{colSpan: 3, renderData: () => <div />}, {renderData: () => null}], {renderData: () => null}
     */
    colSpan?: (record: RowType, rowIndex: number, colIndex: number) => number;
};
