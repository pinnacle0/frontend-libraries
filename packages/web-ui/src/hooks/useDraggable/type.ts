export type Vector = [number, number];

export interface DragState {
    delta: Vector;
    event: MouseEvent;
}

export type DragHandler = (state: DragState) => void;

export interface DragConfig {
    onDragStart?: DragHandler;
    onDrag?: DragHandler;
    onDragEnd?: DragHandler;
}
