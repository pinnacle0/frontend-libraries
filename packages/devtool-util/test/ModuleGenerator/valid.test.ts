import fs from "fs";
import path from "path";

const tmpDirectory = path.join(__dirname, "./__tmp__/valid");
const tmpReduxStatePath = path.join(tmpDirectory, "type/state.ts");
const tmpCommonOldFeatureTypePath = path.join(tmpDirectory, "module/common/old-feature/type.ts");

describe("ModuleGenerator class", () => {
    const fixtures = getFixtures();

    beforeAll(() => {
        if (fs.existsSync(tmpDirectory)) fs.rmSync(tmpDirectory, {recursive: true});
        fixtures.forEach(_ => {
            const fixtureDirectory = path.dirname(_.path);
            if (!fs.existsSync(fixtureDirectory)) fs.mkdirSync(fixtureDirectory, {recursive: true});
            fs.writeFileSync(_.path, _.data, {encoding: "utf8"});
        });
    });

    afterAll(() => {
        // Comment the following line to see the temp files
        // fs.rmSync(tmpDirectory, {recursive: true});
    });

    test("generate files", async () => {
        fixtures.forEach(_ => expect(fs.existsSync(_.path)).toBe(true));

        jest.doMock("yargs", () => ({
            parseSync: () => ({_: {0: "common/new-feature"}}),
        }));
        jest.doMock("../../src/PrettierUtil", () => ({
            PrettierUtil: {format: () => {}},
        }));

        // Note: inline require ModuleGenerator after `jest.doMock` calls to ensure that require hooks are registered (without relying on ts-jest or babel-jest magic)
        const {ModuleGenerator} = jest.requireActual<typeof import("../../src/ModuleGenerator")>("../../src/ModuleGenerator");
        await new ModuleGenerator.Web({
            srcDirectory: tmpDirectory,
        }).run();

        const expectedNewFiles = [
            path.join(tmpDirectory, "module/common/new-feature/Main/index.tsx"),
            path.join(tmpDirectory, "module/common/new-feature/hooks.ts"),
            path.join(tmpDirectory, "module/common/new-feature/index.ts"),
            path.join(tmpDirectory, "module/common/new-feature/type.ts"),
            path.join(tmpDirectory, "module/common/new-feature/module.ts"),
        ];
        // Check if each expected new file exists
        expectedNewFiles.forEach(path => {
            const exists = fs.existsSync(path) ? "exists" : "not-exists";
            expect(`${exists}: ${path}`).toBe(`exists: ${path}`);
        });
        // Create a snapshot for each new file
        expectedNewFiles.concat([tmpReduxStatePath, tmpCommonOldFeatureTypePath]).forEach(path => {
            expect(fs.readFileSync(path, {encoding: "utf8"})).toMatchSnapshot();
        });
    });
});

function getFixtures(): {path: string; data: string}[] {
    return [
        {
            path: tmpReduxStatePath,
            data: `import type {State} from "core-fe";
import type {State as OldFeatureState} from "./modules/common/old-feature";

export interface RootState extends State {
    app: {
        oldFeature: OldFeatureState;
    };
}
`,
        },
        {
            path: tmpCommonOldFeatureTypePath,
            data: `//
export interface State {}
`,
        },
    ];
}
