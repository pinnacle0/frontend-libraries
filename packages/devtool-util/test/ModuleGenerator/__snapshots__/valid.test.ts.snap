// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModuleGenerator class generate files 1`] = `
"import React from \"react\";
import {ReactUtil} from \"@pinnacle0/web-ui/util/ReactUtil\";
import \"./index.less\";

export const Main = ReactUtil.memo(\"Main\", () => {
    return <div />;
});
"
`;

exports[`ModuleGenerator class generate files 2`] = `
"import {useSelector} from \"core-fe\";
import type {RootState} from \"type/state\";

export function useNewFeatureState<T>(fn: (state: RootState[\"app\"][\"newFeature\"]) => T): T {
    return useSelector((state: RootState) => fn(state.app.newFeature));
}
"
`;

exports[`ModuleGenerator class generate files 3`] = `
"import {Module, register} from \"core-fe\";
import {Main} from \"./Main\";
import type {SagaGenerator} from \"core-fe\";
import type {RootState} from \"type/state\";
import type {State} from \"./type\";

const initialState: State = {};

class NewFeatureModule extends Module<RootState, \"newFeature\"> {
    override *onEnter(): SagaGenerator {
        // TODO
    }
}

const newFeatureModule = register(new NewFeatureModule(\"newFeature\", initialState));
export const actions = newFeatureModule.getActions();
export const MainComponent: React.ComponentType<object> = newFeatureModule.attachLifecycle(Main);
"
`;

exports[`ModuleGenerator class generate files 4`] = `
"export interface State {}
"
`;

exports[`ModuleGenerator class generate files 5`] = `
"import type {State} from \"core-fe\";
import type {State as NewFeatureState} from \"module/common/new-feature/type\";
import type {State as OldFeatureState} from \"./modules/common/old-feature\";

export interface RootState extends State {
    app: {
        oldFeature: OldFeatureState;
    newFeature: NewFeatureState;
};
}
"
`;

exports[`ModuleGenerator class generate files 6`] = `
"//
export interface State {}
"
`;
